### ISSUE TITLE
You are passing an undefined module! Please check the object you are passing to i18next.use()

### ISSUE THREAD

**Issue by lucasfelixs** [i18next/react-i18next#1572]
## 🐛 Bug Report

I followed the usage example files i18n with jest. I created the mock file, included in moduleNameMapper, but when I go to test a component that uses i18n, I get this error: `You are passing an undefined module! Please check the object you are passing to i18next.use()`

![Captura de Tela 2022-10-27 às 12 31 14](https://user-images.githubusercontent.com/64923932/198333473-59f68ada-b57f-415f-ad25-c801596a6254.png)

## To Reproduce

`/src/test/__mocks__/react-i18next.js`:
```js
const React = require('react');
const reactI18next = require('react-i18next');

const hasChildren = node => node && (node.children || (node.props && node.props.children));

const getChildren = node =>
  node && node.children ? node.children : node.props && node.props.children;

const renderNodes = reactNodes => {
  if (typeof reactNodes === 'string') {
    return reactNodes;
  }

  return Object.keys(reactNodes).map((key, i) => {
    const child = reactNodes[key];
    const isElement = React.isValidElement(child);

    if (typeof child === 'string') {
      return child;
    }
    if (hasChildren(child)) {
      const inner = renderNodes(getChildren(child));
      return React.cloneElement(child, { ...child.props, key: i }, inner);
    }
    if (typeof child === 'object' && !isElement) {
      return Object.keys(child).reduce((str, childKey) => `${str}${child[childKey]}`, '');
    }

    return child;
  });
};

const useMock = [k => k, {}];
useMock.t = k => k;
useMock.i18n = {};

module.exports = {
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  withTranslation: () => Component => props => <Component t={k => k} {...props} />,
  Trans: ({ children }) =>
    Array.isArray(children) ? renderNodes(children) : renderNodes([children]),
  Translation: ({ children }) => children(k => k, { i18n: {} }),
  useTranslation: () => useMock,

  // mock if needed
  I18nextProvider: reactI18next.I18nextProvider,
  initReactI18next: reactI18next.initReactI18next,
  setDefaults: reactI18next.setDefaults,
  getDefaults: reactI18next.getDefaults,
  setI18n: reactI18next.setI18n,
  getI18n: reactI18next.getI18n,
};
```

`jest.config.js`
```js
moduleNameMapper: {
    'react-i18next': '<rootDir>/src/test/__mocks__/react-i18next.js',
  },
```

## Your Environment

- *node*: v16.14.2
- *i18next version*: ^21.6.16
- *os*: Mac
- *react*: 18.1.0
- *next*: 12.1.6

**Comment by adrai** [comment]
can you console.log initReactI18next

**Comment by adrai** [comment]
And what if you:

```js
import ReactI18next from 'react-i18next';
console.log(ReactI18next)
console.log(ReactI18next.initReactI18next)
```

I suspect your mocking setup has problems mixing cjs exports (module.exports) with esm import ( import { initReactI18next } from 'react-i18next'; )

**Comment by lucasfelixc** [comment]
But do you want to see the console.log in the mocking setup file or in the i18n file config?

**Comment by adrai** [comment]
i18n

**Comment by adrai** [comment]
exactly

**Comment by adrai** [comment]
You need to rewrite this file (/src/test/__mocks__/react-i18next.js) to be an esm module.

**Comment by lucasfelixc** [comment]
Ohh, I got it

**Comment by lucasfelixc** [comment]
I rewrite the `/src/test/mocks/react-i18next.js` like this:

```js
import React from 'react';
import * as reactI18next from 'react-i18next';

const hasChildren = node => node && (node.children || (node.props && node.props.children));

const getChildren = node =>
  node && node.children ? node.children : node.props && node.props.children;

const renderNodes = reactNodes => {
  if (typeof reactNodes === 'string') {
    return reactNodes;
  }

  return Object.keys(reactNodes).map((key, i) => {
    const child = reactNodes[key];
    const isElement = React.isValidElement(child);

    if (typeof child === 'string') {
      return child;
    }
    if (hasChildren(child)) {
      const inner = renderNodes(getChildren(child));
      return React.cloneElement(child, { ...child.props, key: i }, inner);
    }
    if (typeof child === 'object' && !isElement) {
      return Object.keys(child).reduce((str, childKey) => `${str}${child[childKey]}`, '');
    }

    return child;
  });
};

const useMock = [k => k, {}];
useMock.t = k => k;
useMock.i18n = {};

module.exports = {
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  withTranslation: () => Component => props => <Component t={k => k} {...props} />,
  Trans: ({ children }) =>
    Array.isArray(children) ? renderNodes(children) : renderNodes([children]),
  Translation: ({ children }) => children(k => k, { i18n: {} }),
  useTranslation: () => useMock,

  // mock if needed
  I18nextProvider: reactI18next.I18nextProvider,
  initReactI18next: reactI18next.initReactI18next,
  setDefaults: reactI18next.setDefaults,
  getDefaults: reactI18next.getDefaults,
  setI18n: reactI18next.setI18n,
  getI18n: reactI18next.getI18n,
};
```

But I'm getting the same error D:

**Comment by adrai** [comment]
This is not esm export

**Comment by lucasfelixc** [comment]
Oh true, I'm sorry, I got confused

**Comment by lucasfelixc** [comment]
Now with esm export:

```js
import React from 'react';
import { I18nextProvider, initReactI18next, setDefaults, getDefaults, setI18n, getI18n } from 'react-i18next';

const hasChildren = node => node && (node.children || (node.props && node.props.children));

const getChildren = node =>
  node && node.children ? node.children : node.props && node.props.children;

const renderNodes = reactNodes => {
  if (typeof reactNodes === 'string') {
    return reactNodes;
  }

  return Object.keys(reactNodes).map((key, i) => {
    const child = reactNodes[key];
    const isElement = React.isValidElement(child);

    if (typeof child === 'string') {
      return child;
    }
    if (hasChildren(child)) {
      const inner = renderNodes(getChildren(child));
      return React.cloneElement(child, { ...child.props, key: i }, inner);
    }
    if (typeof child === 'object' && !isElement) {
      return Object.keys(child).reduce((str, childKey) => `${str}${child[childKey]}`, '');
    }

    return child;
  });
};

const useMock = [k => k, {}];
useMock.t = k => k;
useMock.i18n = {};

const withTranslation = () => Component => props => <Component t={k => k} {...props} />;
const Trans = ({ children }) =>
    Array.isArray(children) ? renderNodes(children) : renderNodes([children]);
const Translation = ({ children }) => children(k => k, { i18n: {} });
const useTranslation = () => useMock;

export {
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  withTranslation,
  Trans,
  Translation,
  useTranslation,

  // mock if needed
  I18nextProvider,
  initReactI18next,
  setDefaults,
  getDefaults,
  setI18n,
  getI18n,
};
```

I'm getting this error:

![Captura de Tela 2022-10-27 às 13 01 12](https://user-images.githubusercontent.com/64923932/198340763-3918dbe0-e0cd-4fe0-9745-8ca572e82328.png)

**Comment by adrai** [comment]
sorry, I'm afk... not easy to help...
you need to investigate on your own...

**Comment by dashcraft** [comment]
@adrai This ticket, didn't want to post to twitter and i was going to dm you. I skipped the tests that where failing but the mocks are incorrect and don't actually work if you use the useTranslation hook and <Trans> component inside the same page. In my case, i use the Trans Component at runtime in a consts folder to generate static re-usable options dropdowns, i use the translation hook inside of the react component.

TLDR
The mocks don't work as suggested above and it seems related to initReactI18next and usage of hooks + components used in tandem.

**Comment by adrai** [comment]
> @adrai This ticket, didn't want to post to twitter and i was going to dm you. I skipped the tests that where failing but the mocks are incorrect and don't actually work if you use the useTranslation hook and <Trans> component inside the same page. In my case, i use the Trans Component at runtime in a consts folder to generate static re-usable options dropdowns, i use the translation hook inside of the react component.
> 
> 
> 
> TLDR
> 
> The mocks don't work as suggested above and it seems related to initReactI18next and usage of hooks + components used in tandem.

@dashcraft Would be nice to have a reproducible example to see this issues in actions...

**Comment by dashcraft** [comment]
@adrai I understand reproducibility is a core concern for you (although a directly reproducible mock was given above, when dropped and placed in my app broke tests) and I understand your time is limited

However...
I'm not using your package any longer because _it didn't work_ and the suggestions to fix _didn't work_. For me, this is no longer a concern, but I think for your package (which your business seems to use/directly profit from as well) the benefit ratio may be different and it may warrant some of your attention to attempt to at least remotely repro the above example, and dedicate resources to it.

It may not be worth it for you either, that's perfectly fine, and your right as an open-source maintainer.

**Comment by adrai** [comment]
@dashcraft like said, I was and am not able to reproduce that issue in my test setups... that's why I ask for a minimal reproducible example... a small repository or similar. 🤷‍♂️

**Comment by andersonmalheiro** [comment]
I think I arrived a little bit late but @lucasfelixc if you have something like this in your tests:
```
jest.mock('react-i18next', () => ({
    useTranslation: () => ({ t: (key: string | string[]) => key }),
    Trans: ({ i18nKey }: { i18nKey: string }) => i18nKey,
}));

```

you should remove it.
I was having this problem and just realized that was the mock that was breaking the things.

**Comment by tonhao-dev** [comment]
I solve this problem adding `initReactI18next` in my `jest.mock` in top of my suit test file.

```js
jest.mock('react-i18next', () => ({
  useTranslation: jest.fn(),
  initReactI18next: {
    type: '3rdParty',
    init: (i18next) => i18next,
  },
}));
```
